File: app/src/main/java/de/westnordost/osmagent/quests/statistics/QuestStatisticsDao.java
package de.westnordost.osmagent.quests.statistics;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import java.util.HashMap;
import java.util.Map;

import javax.inject.Inject;

import de.westnordost.osmagent.OsmagentConstants;
import de.westnordost.osmagent.quests.QuestType;
import de.westnordost.osmapi.changesets.ChangesetInfo;
import de.westnordost.osmapi.changesets.ChangesetsDao;
import de.westnordost.osmapi.changesets.QueryChangesetsFilters;
import de.westnordost.osmapi.common.Handler;

public class QuestStatisticsDao
{
	private final SQLiteOpenHelper dbHelper;
	private final ChangesetsDao changesetsDao;

	@Inject
	public QuestStatisticsDao(SQLiteOpenHelper dbHelper, ChangesetsDao changesetsDao)
	{
		this.dbHelper = dbHelper;
		this.changesetsDao = changesetsDao;
	}

	public void syncFromOsmServer(long userId)
	{
		QueryChangesetsFilters filters = new QueryChangesetsFilters().byUser(userId).onlyClosed();
		final Map<String, Integer> data = new HashMap<>();

		changesetsDao.find(new Handler<ChangesetInfo>()
		{
			@Override public void handle(ChangesetInfo changeset)
			{
				if(changeset.tags == null) return;

				String userAgent = changeset.tags.get("created_by");
				if(!userAgent.startsWith(OsmagentConstants.NAME)) return;

				String questType = changeset.tags.get(OsmagentConstants.QUESTTYPE_TAG_KEY);
				if(questType == null) return;

				int prev = data.get(questType) != null ? data.get(questType) : 0;
				data.put(questType, prev+1);
			}
		}, filters);

		SQLiteDatabase db = dbHelper.getWritableDatabase();
		// clear table
		db.delete(QuestStatisticsTable.NAME, null, null);

		for(Map.Entry<String, Integer> dataForQuestType : data.entrySet())
		{
			ContentValues values = new ContentValues(2);
			values.put(QuestStatisticsTable.Columns.QUEST_TYPE, dataForQuestType.getKey());
			values.put(QuestStatisticsTable.Columns.SUCCEEDED, dataForQuestType.getValue());
			db.insert(QuestStatisticsTable.NAME, null, values);
		}
	}

	public void addOne(String questType)
	{
		SQLiteDatabase db = dbHelper.getWritableDatabase();

		// first ensure the row exists
		ContentValues values = new ContentValues();
		values.put(QuestStatisticsTable.Columns.QUEST_TYPE, questType);
		values.put(QuestStatisticsTable.Columns.SUCCEEDED, 0);
		db.insertWithOnConflict(QuestStatisticsTable.NAME, null, values, SQLiteDatabase.CONFLICT_IGNORE);

		// then increase by one
		String[] args = {questType};
		db.execSQL("UPDATE " + QuestStatisticsTable.NAME + " SET " +
				QuestStatisticsTable.Columns.SUCCEEDED + " = " + QuestStatisticsTable.Columns.SUCCEEDED +
				" + 1 WHERE " + QuestStatisticsTable.Columns.QUEST_TYPE + " = ?", args);
	}

	public int getAmount(String questType)
	{
		SQLiteDatabase db = dbHelper.getReadableDatabase();

		Cursor cursor = db.query(QuestStatisticsTable.NAME,
				new String[]{QuestStatisticsTable.Columns.SUCCEEDED},
				QuestStatisticsTable.Columns.QUEST_TYPE + " = ?",
				new String[]{questType},
				null, null, null, "1");

		try
		{
			if(!cursor.moveToFirst()) return 0;
			return cursor.getInt(0);
		}
		finally
		{
			cursor.close();
		}
	}

	public int getTotalAmount()
	{
		SQLiteDatabase db = dbHelper.getReadableDatabase();

		String[] cols = {"total("+QuestStatisticsTable.Columns.SUCCEEDED+")"};
		Cursor cursor = db.query(QuestStatisticsTable.NAME, cols, null,	null, null, null, null, null);

		try
		{
			if(!cursor.moveToFirst()) return 0;
			return cursor.getInt(0);
		}
		finally
		{
			cursor.close();
		}
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/data/user/QuestStatisticsDao.kt
package de.westnordost.streetcomplete.data.user

import de.westnordost.streetcomplete.data.Database

import javax.inject.Inject

import de.westnordost.streetcomplete.data.user.QuestStatisticsTable.Columns.QUEST_TYPE
import de.westnordost.streetcomplete.data.user.QuestStatisticsTable.Columns.SUCCEEDED
import de.westnordost.streetcomplete.data.user.QuestStatisticsTable.NAME
import java.util.concurrent.CopyOnWriteArrayList
import javax.inject.Singleton

/** Stores how many quests of which quest types the user solved */
@Singleton class QuestStatisticsDao @Inject constructor(private val db: Database) {

    interface Listener {
        fun onAddedOne(questType: String)
        fun onSubtractedOne(questType: String)
        fun onReplacedAll()
    }

    private val listeners: MutableList<Listener> = CopyOnWriteArrayList()

    fun getTotalAmount(): Int =
        db.queryOne(NAME, arrayOf("total($SUCCEEDED) as count")) { it.getInt("count") } ?: 0

    fun getAll(): Map<String, Int> =
        db.query(NAME) { it.getString(QUEST_TYPE) to it.getInt(SUCCEEDED) }.toMap()

    fun clear() {
        db.delete(NAME)
        listeners.forEach { it.onReplacedAll() }
    }

    fun replaceAll(amounts: Map<String, Int>) {
        db.transaction {
            db.delete(NAME)
            if (amounts.isNotEmpty()) {
                db.replaceMany(NAME,
                    arrayOf(QUEST_TYPE, SUCCEEDED),
                    amounts.map { arrayOf(it.key, it.value) }
                )
            }
        }
        listeners.forEach { it.onReplacedAll() }
    }

    fun addOne(questType: String) {
        db.transaction {
            // first ensure the row exists
            db.insertOrIgnore(NAME, listOf(
                QUEST_TYPE to questType,
                SUCCEEDED to 0
            ))

            // then increase by one
            db.exec("UPDATE $NAME SET $SUCCEEDED = $SUCCEEDED + 1 WHERE $QUEST_TYPE = ?", arrayOf(questType))
        }

        listeners.forEach { it.onAddedOne(questType) }
    }

    fun subtractOne(questType: String) {
        db.exec("UPDATE $NAME SET $SUCCEEDED = $SUCCEEDED - 1 WHERE $QUEST_TYPE = ?", arrayOf(questType))
        listeners.forEach { it.onSubtractedOne(questType) }
    }

    fun getAmount(questType: String): Int =
        db.queryOne(NAME,
            columns = arrayOf(SUCCEEDED),
            where = "$QUEST_TYPE = ?",
            args = arrayOf(questType)
        ) { it.getInt(SUCCEEDED) } ?: 0

    fun getAmount(questTypes: List<String>): Int {
        val questionMarks = Array(questTypes.size) { "?" }.joinToString(",")
        return db.queryOne(NAME,
            columns = arrayOf("total($SUCCEEDED) as count"),
            where = "$QUEST_TYPE in ($questionMarks)",
            args = questTypes.toTypedArray()
        ) { it.getInt("count") } ?: 0
    }

    fun addListener(listener: Listener) {
        listeners.add(listener)
    }
    fun removeListener(listener: Listener) {
        listeners.remove(listener)
    }
}
